# EPL DATA ANALYSIS

## Introduction 

The purpose of this notebook is to analyze the impact of covid-19 to the performance of English Premier League (EPL) teams. The dataset contained match information from : 
* 2018/2019 season (1 season before covid-19 interruption)
* 2019/2020 season (season where covid interrupted the league)
* 2020/2021 season (1 season after covid-19 interruption)

Information regarding the date of covid-19 interruption : 
* last match before covid-19 interruption = Leicester vs Aston Villa 10 March 2020 (match ID 46889)
* first match before covid-19 interruption = Aston Villa vs Sheffield United 18 June 2020 (match ID 46875)

Analysis objectives : 
* How the absence of spectators had impacted performance of EPL teams 
    * Home/Away Possession
    * Home/Away Shots on Target
    * Home/Away Shots
    * Home/Away Goals Scored per Game
    * Home/Away Win Rate
    * Home/Away Shot on Target % = Shots on Target / Shots
    * Home/Away Quantity Conversion Rate = Goals / Shots 
    * Home/Away Quality Conversion Rate  = Goals / Shots on Target
    * Home/Away Win Rate = No. of win / No. of games (Home/Away)
    * Home/Away Points per Game (PpG) = Average points/game (Home / Away)
* Relation of Ball Possession % vs Goal Conversion Rate
* Compare the performance metrics between "big-6" and "non big-6" teams before and after covid-19. (note : Big-6 = "Manchester United", "Manchester City", "Liverpool", "Arsenal", "Chelsea", "Tottenham Hotspurs")

In [14]:
import pandas as pd

#Set float number just have two decimals
pd.set_option('display.float_format','{:.2f}' .format)

In [15]:
epl_data = pd.read_csv('epl_data_cleaned.csv', index_col=[0])
epl_data

match_id  match_date     matchweek                home_team  \
0        38309  2018-08-11   Matchweek 1          AFC Bournemouth   
1        38310  2018-08-11   Matchweek 1                   Fulham   
2        38311  2018-08-11   Matchweek 1        Huddersfield Town   
3        38313  2018-08-11   Matchweek 1        Manchester United   
4        38314  2018-08-11   Matchweek 1         Newcastle United   
...        ...         ...           ...                      ...   
1135     59272  2021-05-23  Matchweek 38          Manchester City   
1136     59273  2021-05-23  Matchweek 38         Sheffield United   
1137     59274  2021-05-23  Matchweek 38          West Ham United   
1138     59275  2021-05-23  Matchweek 38  Wolverhampton Wanderers   
1139     59266  2021-05-23  Matchweek 38                  Arsenal   

                     away_team     season  home_score  away_score  \
0                 Cardiff City  2018/2019           2           0   
1               Crystal Palace  2018/2019           0           2   
2                      Chelsea  2018/2019           0           3   
3               Leicester City  2018/2019           2           1   
4            Tottenham Hotspur  2018/2019           1           2   
...                        ...        ...         ...         ...   
1135                   Everton  2020/2021           5           0   
1136                   Burnley  2020/2021           1           0   
1137               Southampton  2020/2021           3           0   
1138         Manchester United  2020/2021           1           2   
1139  Brighton and Hove Albion  2020/2021           2           0   

      home_possession  away_possession  home_shots_on_target  \
0               62.90            37.10                     4   
1               66.30            33.70                     6   
2               37.20            62.80                     1   
3               46.30            53.70                     6   
4               40.40            59.60                     2   
...               ...              ...                   ...   
1135            67.70            32.30                    11   
1136            43.00            57.00                     3   
1137            36.90            63.10                     7   
1138            57.20            42.80                     4   
1139            53.90            46.10                     5   

      away_shots_on_target  home_shots  away_shots  home_points  away_points  
0                        1          12          10            3            0  
1                       10          15          12            0            3  
2                        4           6          13            0            3  
3                        4           8          13            3            0  
4                        5          15          15            0            3  
...                    ...         ...         ...          ...          ...  
1135                     3          21           8            3            0  
1136                     3          12          10            3            0  
1137                     5          14          17            3            0  
1138                     4          14           9            0            3  
1139                     1          16           5            3            0  

[1140 rows x 16 columns]

## Calculate Key Stats 

Some of the key stats required for further analysis are : 
* Shot on Target % = Shots on Target / Shots
* Quantity Conversion Rate = Goals / Shots 
* Quality Conversion Rate  = Goals / Shots on Target

In [16]:
def calculate_key_stats(df, num, div,stats_name,percentage=True):
    if percentage == True:
        #home key stats
        df['home_'+stats_name] = (df['home_'+num]/df['home_'+div]) * 100
        #away key stats
        df['away_'+stats_name] = (df['away_'+num]/df['away_'+div]) * 100
    else:
        #home key stats
        df['home_'+stats_name] = df['home_'+num]/df['home_'+div]
        #away key stats
        df['away_'+stats_name] = df['away_'+num]/df['away_'+div]
    
    #handling 0/0 value (if any)
    df = df.fillna(0, axis=1)
    
    return df

In [17]:
#calculate shot on target %
calculate_key_stats(df=epl_data, num='shots_on_target',div='shots',stats_name='SoT%')

#calculate quantity conversion rate
calculate_key_stats(df=epl_data, num='score',div='shots',stats_name='quan_CR')

#calculate quality conversion rate
calculate_key_stats(df=epl_data, num='score',div='shots_on_target',stats_name='qual_CR')

match_id  match_date     matchweek                home_team  \
0        38309  2018-08-11   Matchweek 1          AFC Bournemouth   
1        38310  2018-08-11   Matchweek 1                   Fulham   
2        38311  2018-08-11   Matchweek 1        Huddersfield Town   
3        38313  2018-08-11   Matchweek 1        Manchester United   
4        38314  2018-08-11   Matchweek 1         Newcastle United   
...        ...         ...           ...                      ...   
1135     59272  2021-05-23  Matchweek 38          Manchester City   
1136     59273  2021-05-23  Matchweek 38         Sheffield United   
1137     59274  2021-05-23  Matchweek 38          West Ham United   
1138     59275  2021-05-23  Matchweek 38  Wolverhampton Wanderers   
1139     59266  2021-05-23  Matchweek 38                  Arsenal   

                     away_team     season  home_score  away_score  \
0                 Cardiff City  2018/2019           2           0   
1               Crystal Palace  2018/2019           0           2   
2                      Chelsea  2018/2019           0           3   
3               Leicester City  2018/2019           2           1   
4            Tottenham Hotspur  2018/2019           1           2   
...                        ...        ...         ...         ...   
1135                   Everton  2020/2021           5           0   
1136                   Burnley  2020/2021           1           0   
1137               Southampton  2020/2021           3           0   
1138         Manchester United  2020/2021           1           2   
1139  Brighton and Hove Albion  2020/2021           2           0   

      home_possession  away_possession  ...  home_shots  away_shots  \
0               62.90            37.10  ...          12          10   
1               66.30            33.70  ...          15          12   
2               37.20            62.80  ...           6          13   
3               46.30            53.70  ...           8          13   
4               40.40            59.60  ...          15          15   
...               ...              ...  ...         ...         ...   
1135            67.70            32.30  ...          21           8   
1136            43.00            57.00  ...          12          10   
1137            36.90            63.10  ...          14          17   
1138            57.20            42.80  ...          14           9   
1139            53.90            46.10  ...          16           5   

      home_points  away_points  home_SoT%  away_SoT%  home_quan_CR  \
0               3            0      33.33      10.00         16.67   
1               0            3      40.00      83.33          0.00   
2               0            3      16.67      30.77          0.00   
3               3            0      75.00      30.77         25.00   
4               0            3      13.33      33.33          6.67   
...           ...          ...        ...        ...           ...   
1135            3            0      52.38      37.50         23.81   
1136            3            0      25.00      30.00          8.33   
1137            3            0      50.00      29.41         21.43   
1138            0            3      28.57      44.44          7.14   
1139            3            0      31.25      20.00         12.50   

      away_quan_CR  home_qual_CR  away_qual_CR  
0             0.00         50.00          0.00  
1            16.67          0.00         20.00  
2            23.08          0.00         75.00  
3             7.69         33.33         25.00  
4            13.33         50.00         40.00  
...            ...           ...           ...  
1135          0.00         45.45          0.00  
1136          0.00         33.33          0.00  
1137          0.00         42.86          0.00  
1138         22.22         25.00         50.00  
1139          0.00         40.00          0.00  

[1140 rows x 22 columns]

## Calculate Points

In [18]:
def get_points(df):
    #home win
    df.loc[df['home_score']>df['away_score'], 'home_points'] = 3
    df.loc[df['home_score']>df['away_score'], 'away_points'] = 0

    #away win
    df.loc[df['home_score']<df['away_score'], 'home_points'] = 0
    df.loc[df['home_score']<df['away_score'], 'away_points'] = 3
    
    #draw
    df.loc[df['home_score']==df['away_score'], 'home_points'] = 1
    df.loc[df['home_score']==df['away_score'], 'away_points'] = 1
    
    return df
        
#calculate points for each game
get_points(epl_data)

match_id  match_date     matchweek                home_team  \
0        38309  2018-08-11   Matchweek 1          AFC Bournemouth   
1        38310  2018-08-11   Matchweek 1                   Fulham   
2        38311  2018-08-11   Matchweek 1        Huddersfield Town   
3        38313  2018-08-11   Matchweek 1        Manchester United   
4        38314  2018-08-11   Matchweek 1         Newcastle United   
...        ...         ...           ...                      ...   
1135     59272  2021-05-23  Matchweek 38          Manchester City   
1136     59273  2021-05-23  Matchweek 38         Sheffield United   
1137     59274  2021-05-23  Matchweek 38          West Ham United   
1138     59275  2021-05-23  Matchweek 38  Wolverhampton Wanderers   
1139     59266  2021-05-23  Matchweek 38                  Arsenal   

                     away_team     season  home_score  away_score  \
0                 Cardiff City  2018/2019           2           0   
1               Crystal Palace  2018/2019           0           2   
2                      Chelsea  2018/2019           0           3   
3               Leicester City  2018/2019           2           1   
4            Tottenham Hotspur  2018/2019           1           2   
...                        ...        ...         ...         ...   
1135                   Everton  2020/2021           5           0   
1136                   Burnley  2020/2021           1           0   
1137               Southampton  2020/2021           3           0   
1138         Manchester United  2020/2021           1           2   
1139  Brighton and Hove Albion  2020/2021           2           0   

      home_possession  away_possession  ...  home_shots  away_shots  \
0               62.90            37.10  ...          12          10   
1               66.30            33.70  ...          15          12   
2               37.20            62.80  ...           6          13   
3               46.30            53.70  ...           8          13   
4               40.40            59.60  ...          15          15   
...               ...              ...  ...         ...         ...   
1135            67.70            32.30  ...          21           8   
1136            43.00            57.00  ...          12          10   
1137            36.90            63.10  ...          14          17   
1138            57.20            42.80  ...          14           9   
1139            53.90            46.10  ...          16           5   

      home_points  away_points  home_SoT%  away_SoT%  home_quan_CR  \
0               3            0      33.33      10.00         16.67   
1               0            3      40.00      83.33          0.00   
2               0            3      16.67      30.77          0.00   
3               3            0      75.00      30.77         25.00   
4               0            3      13.33      33.33          6.67   
...           ...          ...        ...        ...           ...   
1135            3            0      52.38      37.50         23.81   
1136            3            0      25.00      30.00          8.33   
1137            3            0      50.00      29.41         21.43   
1138            0            3      28.57      44.44          7.14   
1139            3            0      31.25      20.00         12.50   

      away_quan_CR  home_qual_CR  away_qual_CR  
0             0.00         50.00          0.00  
1            16.67          0.00         20.00  
2            23.08          0.00         75.00  
3             7.69         33.33         25.00  
4            13.33         50.00         40.00  
...            ...           ...           ...  
1135          0.00         45.45          0.00  
1136          0.00         33.33          0.00  
1137          0.00         42.86          0.00  
1138         22.22         25.00         50.00  
1139          0.00         40.00          0.00  

[1140 rows x 22 columns]

## Calculate Win Rate

In [19]:
def get_win_rate(df):
    #home win
    df.loc[df['home_score']>df['away_score'], 'home_win_rate'] = 100
    df.loc[df['home_score']>df['away_score'], 'away_win_rate'] = 0

    #away win
    df.loc[df['home_score']<df['away_score'], 'home_win_rate'] = 0
    df.loc[df['home_score']<df['away_score'], 'away_win_rate'] = 100
    
    #draw
    df.loc[df['home_score']==df['away_score'], 'home_win_rate'] = 0
    df.loc[df['home_score']==df['away_score'], 'away_win_rate'] = 0
    
    return df

#calculate win rate
get_win_rate(epl_data)

match_id  match_date     matchweek                home_team  \
0        38309  2018-08-11   Matchweek 1          AFC Bournemouth   
1        38310  2018-08-11   Matchweek 1                   Fulham   
2        38311  2018-08-11   Matchweek 1        Huddersfield Town   
3        38313  2018-08-11   Matchweek 1        Manchester United   
4        38314  2018-08-11   Matchweek 1         Newcastle United   
...        ...         ...           ...                      ...   
1135     59272  2021-05-23  Matchweek 38          Manchester City   
1136     59273  2021-05-23  Matchweek 38         Sheffield United   
1137     59274  2021-05-23  Matchweek 38          West Ham United   
1138     59275  2021-05-23  Matchweek 38  Wolverhampton Wanderers   
1139     59266  2021-05-23  Matchweek 38                  Arsenal   

                     away_team     season  home_score  away_score  \
0                 Cardiff City  2018/2019           2           0   
1               Crystal Palace  2018/2019           0           2   
2                      Chelsea  2018/2019           0           3   
3               Leicester City  2018/2019           2           1   
4            Tottenham Hotspur  2018/2019           1           2   
...                        ...        ...         ...         ...   
1135                   Everton  2020/2021           5           0   
1136                   Burnley  2020/2021           1           0   
1137               Southampton  2020/2021           3           0   
1138         Manchester United  2020/2021           1           2   
1139  Brighton and Hove Albion  2020/2021           2           0   

      home_possession  away_possession  ...  home_points  away_points  \
0               62.90            37.10  ...            3            0   
1               66.30            33.70  ...            0            3   
2               37.20            62.80  ...            0            3   
3               46.30            53.70  ...            3            0   
4               40.40            59.60  ...            0            3   
...               ...              ...  ...          ...          ...   
1135            67.70            32.30  ...            3            0   
1136            43.00            57.00  ...            3            0   
1137            36.90            63.10  ...            3            0   
1138            57.20            42.80  ...            0            3   
1139            53.90            46.10  ...            3            0   

      home_SoT%  away_SoT%  home_quan_CR  away_quan_CR  home_qual_CR  \
0         33.33      10.00         16.67          0.00         50.00   
1         40.00      83.33          0.00         16.67          0.00   
2         16.67      30.77          0.00         23.08          0.00   
3         75.00      30.77         25.00          7.69         33.33   
4         13.33      33.33          6.67         13.33         50.00   
...         ...        ...           ...           ...           ...   
1135      52.38      37.50         23.81          0.00         45.45   
1136      25.00      30.00          8.33          0.00         33.33   
1137      50.00      29.41         21.43          0.00         42.86   
1138      28.57      44.44          7.14         22.22         25.00   
1139      31.25      20.00         12.50          0.00         40.00   

      away_qual_CR  home_win_rate  away_win_rate  
0             0.00         100.00           0.00  
1            20.00           0.00         100.00  
2            75.00           0.00         100.00  
3            25.00         100.00           0.00  
4            40.00           0.00         100.00  
...            ...            ...            ...  
1135          0.00         100.00           0.00  
1136          0.00         100.00           0.00  
1137          0.00         100.00           0.00  
1138         50.00           0.00         100.00  
1139          0.00         100.00           0

## Change Datatype

In [20]:
epl_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1140 entries, 0 to 1139
Data columns (total 24 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   match_id              1140 non-null   int64  
 1   match_date            1140 non-null   object 
 2   matchweek             1140 non-null   object 
 3   home_team             1140 non-null   object 
 4   away_team             1140 non-null   object 
 5   season                1140 non-null   object 
 6   home_score            1140 non-null   int64  
 7   away_score            1140 non-null   int64  
 8   home_possession       1140 non-null   float64
 9   away_possession       1140 non-null   float64
 10  home_shots_on_target  1140 non-null   int64  
 11  away_shots_on_target  1140 non-null   int64  
 12  home_shots            1140 non-null   int64  
 13  away_shots            1140 non-null   int64  
 14  home_points           1140 non-null   int64  
 15  away_points          

In [21]:
string_cols = ['matchweek','home_team','away_team','season']
int_cols = ['home_score','away_score','home_shots_on_target','away_shots_on_target',
            'home_shots','away_shots','home_points','away_points']
float_cols = ['home_possession','away_possession','home_SoT%','away_SoT%','home_quan_CR','away_quan_CR' \
              ,'home_qual_CR','away_qual_CR','home_win_rate','away_win_rate']
datetime_cols = 'match_date'

#convert to corresponding datatype
epl_data[string_cols] = epl_data[string_cols].astype("string")
epl_data[int_cols] = epl_data[int_cols].astype("int")
epl_data[float_cols] = epl_data[float_cols].astype("float")

#convert object to datetime object
epl_data['match_date'] = pd.to_datetime(epl_data['match_date'], format='%Y-%m-%d')

epl_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1140 entries, 0 to 1139
Data columns (total 24 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   match_id              1140 non-null   int64         
 1   match_date            1140 non-null   datetime64[ns]
 2   matchweek             1140 non-null   string        
 3   home_team             1140 non-null   string        
 4   away_team             1140 non-null   string        
 5   season                1140 non-null   string        
 6   home_score            1140 non-null   int32         
 7   away_score            1140 non-null   int32         
 8   home_possession       1140 non-null   float64       
 9   away_possession       1140 non-null   float64       
 10  home_shots_on_target  1140 non-null   int32         
 11  away_shots_on_target  1140 non-null   int32         
 12  home_shots            1140 non-null   int32         
 13  away_shots        

In [22]:
#label match before and after covid19
epl_data.loc[epl_data['match_date']<='2020-03-10','covid19'] = 'before'
epl_data.loc[epl_data['match_date']>'2020-03-10','covid19'] = 'after'

epl_data

match_id match_date     matchweek                home_team  \
0        38309 2018-08-11   Matchweek 1          AFC Bournemouth   
1        38310 2018-08-11   Matchweek 1                   Fulham   
2        38311 2018-08-11   Matchweek 1        Huddersfield Town   
3        38313 2018-08-11   Matchweek 1        Manchester United   
4        38314 2018-08-11   Matchweek 1         Newcastle United   
...        ...        ...           ...                      ...   
1135     59272 2021-05-23  Matchweek 38          Manchester City   
1136     59273 2021-05-23  Matchweek 38         Sheffield United   
1137     59274 2021-05-23  Matchweek 38          West Ham United   
1138     59275 2021-05-23  Matchweek 38  Wolverhampton Wanderers   
1139     59266 2021-05-23  Matchweek 38                  Arsenal   

                     away_team     season  home_score  away_score  \
0                 Cardiff City  2018/2019           2           0   
1               Crystal Palace  2018/2019           0           2   
2                      Chelsea  2018/2019           0           3   
3               Leicester City  2018/2019           2           1   
4            Tottenham Hotspur  2018/2019           1           2   
...                        ...        ...         ...         ...   
1135                   Everton  2020/2021           5           0   
1136                   Burnley  2020/2021           1           0   
1137               Southampton  2020/2021           3           0   
1138         Manchester United  2020/2021           1           2   
1139  Brighton and Hove Albion  2020/2021           2           0   

      home_possession  away_possession  ...  away_points  home_SoT%  \
0               62.90            37.10  ...            0      33.33   
1               66.30            33.70  ...            3      40.00   
2               37.20            62.80  ...            3      16.67   
3               46.30            53.70  ...            0      75.00   
4               40.40            59.60  ...            3      13.33   
...               ...              ...  ...          ...        ...   
1135            67.70            32.30  ...            0      52.38   
1136            43.00            57.00  ...            0      25.00   
1137            36.90            63.10  ...            0      50.00   
1138            57.20            42.80  ...            3      28.57   
1139            53.90            46.10  ...            0      31.25   

      away_SoT%  home_quan_CR  away_quan_CR  home_qual_CR  away_qual_CR  \
0         10.00         16.67          0.00         50.00          0.00   
1         83.33          0.00         16.67          0.00         20.00   
2         30.77          0.00         23.08          0.00         75.00   
3         30.77         25.00          7.69         33.33         25.00   
4         33.33          6.67         13.33         50.00         40.00   
...         ...           ...           ...           ...           ...   
1135      37.50         23.81          0.00         45.45          0.00   
1136      30.00          8.33          0.00         33.33          0.00   
1137      29.41         21.43          0.00         42.86          0.00   
1138      44.44          7.14         22.22         25.00         50.00   
1139      20.00         12.50          0.00         40.00          0.00   

      home_win_rate  away_win_rate  covid19  
0            100.00           0.00   before  
1              0.00         100.00   before  
2              0.00         100.00   before  
3            100.00           0.00   before  
4              0.00         100.00   before  
...             ...            ...      ...  
1135         100.00           0.00    after  
1136         100.00           0.00    after  
1137         100.00           0.00    after  
1138           0.00         100.00    after  
1139         100.00           0.00    after  

[1140 rows x 25 columns]

## Filter EPL Teams

The analysis will only observe the teams which had participated in all 3 seasons.

In [23]:
#get teams name for every season since 2018/2019 
teams_18_19 = set(epl_data.loc[epl_data['season']=='2018/2019']['home_team'].unique())
teams_19_20 = set(epl_data.loc[epl_data['season']=='2019/2020']['home_team'].unique())
teams_20_21 = set(epl_data.loc[epl_data['season']=='2020/2021']['home_team'].unique())

#filter the team that had participated in all 3 seasons
filtered_teams = list(teams_18_19.intersection(teams_19_20,teams_20_21))
filtered_teams

['Crystal Palace',
 'Chelsea',
 'Manchester City',
 'Burnley',
 'Manchester United',
 'Everton',
 'Liverpool',
 'Wolverhampton Wanderers',
 'Tottenham Hotspur',
 'Newcastle United',
 'Southampton',
 'Arsenal',
 'Brighton and Hove Albion',
 'West Ham United',
 'Leicester City']

## Data Analysis

The changes in stats are evaluated in 3 categories : 
* If the stats increases 5 % or more, it is categorized as 'positive effect'
* If the stats decreases -5 % or more, it is categorized as 'negative effect'
* If the stats changes within range -5% to 5%, it is categorized as 'relatively not affected'

In [42]:
def evaluate_stats(df,pos_chg=True,pct=True,chg_range=5):
    #evaluate changes 
    ev = df.copy()

    if pos_chg == True:
    
        ev.loc[ev['changes_%']>=chg_range,'effect'] = 'positive effect'  
        ev.loc[ev['changes_%']<=-chg_range,'effect'] = 'negative effect'   
        ev.loc[(ev['changes_%']<chg_range)&(ev['changes_%']>-chg_range),'effect'] = 'relatively not affected'  
    
    else:
        
        ev.loc[ev['changes_%']>=chg_range,'effect'] = 'negative effect'  
        ev.loc[ev['changes_%']<=-chg_range,'effect'] = 'positive effect'   
        ev.loc[(ev['changes_%']<chg_range)&(ev['changes_%']>-chg_range),'effect'] = 'relatively not affected'
            
    evaluation = ev.value_counts('effect').to_frame(name='count')
    
    #Set the order of display
    custom_dict = {'positive effect': 0, 
                   'relatively not affected': 1, 
                   'negative effect': 2} 
    
    evaluation = evaluation.sort_index(key=lambda x:x.map(custom_dict))
    
    return evaluation

In [43]:
def get_home_stats(stats_name,df=epl_data,team=filtered_teams,evaluate=True,positive_changes=True):
    
    print('The effect of covid-19 on EPL home teams ' + stats_name)
    
    home_stats = pd.DataFrame()
    
    #home stats before covid-19
    home_stats['home_'+stats_name+'_before'] = \
        df.loc[df['covid19']=='before'].groupby('home_team')['home_'+stats_name].mean()
    
    #home stats after covid-19
    home_stats['home_'+stats_name+'_after'] = \
        df.loc[df['covid19']=='after'].groupby('home_team')['home_'+stats_name].mean()
    
    #filter team 
    home_stats = home_stats.loc[team]
    
    #changes in home stats
    home_stats['stats_changes'] = home_stats['home_'+stats_name+'_after'] - home_stats['home_'+stats_name+'_before']
    home_stats['changes_%'] = (home_stats['stats_changes']/home_stats['home_'+stats_name+'_before'])*100
    
    #sort by biggest change ascending
    home_stats = home_stats.sort_values('stats_changes', ascending=True)
    
    if evaluate == True:
        display(evaluate_stats(home_stats,pos_chg=positive_changes))
    
    #calculate average home stats before covid
    home_stats_before = home_stats['home_'+stats_name+'_before'].mean()
    print('The average home {} before covid19 is {}'.format(stats_name, round(home_stats_before,2)))
    
    #calculate average home stats after covid
    home_stats_after = home_stats['home_'+stats_name+'_after'].mean()
    print('The average home {} after covid19 is {}'.format(stats_name, round(home_stats_after,2)))
    
    #calculate average home stats changes 
    home_stats_changes = home_stats_after - home_stats_before
    print('The average stats changes of home {} is {}'.format(stats_name, round(home_stats_changes,2)))
    
    #calculate average changes percentage 
    changes_pct = (home_stats_changes/home_stats_before)*100
    print('The average changes % of home {} is {} %'.format(stats_name, round(changes_pct,2)))
    
    return home_stats

In [44]:
def get_away_stats(stats_name,df=epl_data,team=filtered_teams,evaluate=True,percentage=True,positive_changes=True):
    
    print('The effect of covid-19 on EPL away teams ' + stats_name)
    
    away_stats = pd.DataFrame()
    
    #away stats before covid-19
    away_stats['away_'+stats_name+'_before'] = \
        df.loc[df['covid19']=='before'].groupby('away_team')['away_'+stats_name].mean()
    
    #away stats after covid-19
    away_stats['away_'+stats_name+'_after'] = \
        df.loc[df['covid19']=='after'].groupby('away_team')['away_'+stats_name].mean()
    
    #filter team 
    away_stats = away_stats.loc[team]
    
    #changes in home stats
    away_stats['stats_changes'] = away_stats['away_'+stats_name+'_after'] - away_stats['away_'+stats_name+'_before']
    away_stats['changes_%'] = (away_stats['stats_changes']/away_stats['away_'+stats_name+'_before'])*100
    
    #sort by biggest change ascending
    away_stats = away_stats.sort_values('stats_changes', ascending=True)
    
    if evaluate == True:
        display(evaluate_stats(away_stats,pos_chg=positive_changes)) 
    
    #calculate average away stats before covid
    away_stats_before = away_stats['away_'+stats_name+'_before'].mean()
    print('The average away {} before covid19 is {}'.format(stats_name, round(away_stats_before,2)))
    
    #calculate average away stats after covid
    away_stats_after = away_stats['away_'+stats_name+'_after'].mean()
    print('The average away {} after covid19 is {}'.format(stats_name, round(away_stats_after,2)))
    
    #calculate average away stats changes 
    away_stats_changes = away_stats_after - away_stats_before
    print('The average stats changes of away {} is {}'.format(stats_name, round(away_stats_changes,2)))
    
    #calculate average changes percentage 
    changes_pct = (away_stats_changes/away_stats_before)*100
    print('The average changes % of away {} is {} %'.format(stats_name, round(changes_pct,2)))
    
    return away_stats

We can differentiate the stats into 2 groups : 
* Result Stats 
    * win rate
    * points per game
    * goals scored per game
* Performance Stats
    * Ball Possession
    * Shots
    * Shots on Target
    * Shots on Target %
    * Quantity conversion rate
    * Quality conversion rate

### Results Stats

##### Home/Away Win Rate

In [45]:
get_home_stats('win_rate')

The effect of covid-19 on EPL home teams win_rate


count
effect                        
positive effect              2
relatively not affected      4
negative effect              9

The average home win_rate before covid19 is 50.96
The average home win_rate after covid19 is 43.65
The average stats changes of home win_rate is -7.31
The average changes % of home win_rate is -14.34 %


home_win_rate_before  home_win_rate_after  \
home_team                                                             
Liverpool                                94.12                56.52   
Everton                                  51.52                29.17   
Burnley                                  41.18                21.74   
Arsenal                                  61.76                47.83   
Newcastle United                         39.39                29.17   
Brighton and Hove Albion                 30.30                20.83   
Manchester City                          84.38                76.00   
Crystal Palace                           32.35                26.09   
Manchester United                        52.94                47.83   
Tottenham Hotspur                        60.61                58.33   
Leicester City                           50.00                47.83   
Wolverhampton Wanderers                  44.12                43.48   
Chelsea                                  55.88                56.52   
West Ham United                          39.39                50.00   
Southampton                              26.47                43.48   

                          stats_changes  changes_%  
home_team                                           
Liverpool                        -37.60     -39.95  
Everton                          -22.35     -43.38  
Burnley                          -19.44     -47.20  
Arsenal                          -13.94     -22.57  
Newcastle United                 -10.23     -25.96  
Brighton and Hove Albion          -9.47     -31.25  
Manchester City                   -8.38      -9.93  
Crystal Palace                    -6.27     -19.37  
Manchester United                 -5.12      -9.66  
Tottenham Hotspur                 -2.27      -3.75  
Leicester City                    -2.17      -4.35  
Wolverhampton Wanderers           -0.64      -1.45  
Chelsea                            0.64       1.14  
West Ham United                   10.61      26.92  
Southampton                       17.01      64.25

In [46]:
get_away_stats('win_rate')

The effect of covid-19 on EPL away teams win_rate


count
effect                        
positive effect              8
relatively not affected      4
negative effect              3

The average away win_rate before covid19 is 37.23
The average away win_rate after covid19 is 43.67
The average stats changes of away win_rate is 6.44
The average changes % of away win_rate is 17.29 %


away_win_rate_before  away_win_rate_after  \
away_team                                                             
Liverpool                                75.76                50.00   
Crystal Palace                           39.39                29.17   
Wolverhampton Wanderers                  33.33                29.17   
Tottenham Hotspur                        41.18                39.13   
Southampton                              30.30                29.17   
Chelsea                                  48.48                50.00   
Manchester City                          67.65                69.57   
Leicester City                           42.42                45.83   
Newcastle United                         23.53                30.43   
Burnley                                  24.24                37.50   
Brighton and Hove Albion                 14.71                30.43   
West Ham United                          26.47                43.48   
Arsenal                                  28.12                48.00   
Manchester United                        39.39                66.67   
Everton                                  23.53                56.52   

                          stats_changes  changes_%  
away_team                                           
Liverpool                        -25.76     -34.00  
Crystal Palace                   -10.23     -25.96  
Wolverhampton Wanderers           -4.17     -12.50  
Tottenham Hotspur                 -2.05      -4.97  
Southampton                       -1.14      -3.75  
Chelsea                            1.52       3.13  
Manchester City                    1.92       2.84  
Leicester City                     3.41       8.04  
Newcastle United                   6.91      29.35  
Burnley                           13.26      54.69  
Brighton and Hove Albion          15.73     106.96  
West Ham United                   17.01      64.25  
Arsenal                           19.88      70.67  
Manchester United                 27.27      69.23  
Everton                           32.99     140.22

Insight from Home/Away win rate : 
* Generally, covid-19 had negatively affected home win rate and positively affected away win rate, with 9 out of 15 teams had significantly decreased home win rate & 8 out of 15 teams had significantly increased away win rate.
* Southampton is the team with biggest home win rate increase after home crowd absence.
* West Ham is the only team that had improved in both home/away win rate with +0.11 and +0.17 in stats changes respectively.
* Everton experienced the most drastically home/away win rate changes with -0.22 and +0.33 in stats chages respectively. It is also the team with most improved away win rate.
* Liverpool is the most negatively affected team in both home/away win rate with -0.38 and -0.26 in stats changes respectively.
* Chelsea and Tottenham Hotspur are the only teams that is not relatively not affected in both home/away win rate.

#### Home/Away Points per Game

In [47]:
get_home_stats('points')

The effect of covid-19 on EPL home teams points


count
effect                        
positive effect              2
relatively not affected      2
negative effect             11

The average home points before covid19 is 1.74
The average home points after covid19 is 1.54
The average stats changes of home points is -0.21
The average changes % of home points is -11.92 %


home_points_before  home_points_after  \
home_team                                                         
Liverpool                               2.88               1.87   
Everton                                 1.79               1.17   
Arsenal                                 2.09               1.65   
Burnley                                 1.35               1.00   
Manchester City                         2.59               2.36   
Newcastle United                        1.39               1.17   
Manchester United                       1.91               1.70   
Brighton and Hove Albion                1.24               1.04   
Crystal Palace                          1.24               1.04   
Wolverhampton Wanderers                 1.65               1.48   
Leicester City                          1.68               1.52   
Tottenham Hotspur                       1.94               1.92   
Chelsea                                 1.94               1.96   
West Ham United                         1.39               1.71   
Southampton                             1.09               1.48   

                          stats_changes  changes_%  
home_team                                           
Liverpool                         -1.01     -35.14  
Everton                           -0.62     -34.75  
Arsenal                           -0.44     -20.88  
Burnley                           -0.35     -26.09  
Manchester City                   -0.23      -9.01  
Newcastle United                  -0.23     -16.30  
Manchester United                 -0.22     -11.30  
Brighton and Hove Albion          -0.20     -16.16  
Crystal Palace                    -0.19     -15.53  
Wolverhampton Wanderers           -0.17     -10.25  
Leicester City                    -0.15      -9.23  
Tottenham Hotspur                 -0.02      -1.17  
Chelsea                            0.02       0.79  
West Ham United                    0.31      22.55  
Southampton                        0.39      35.84

In [48]:
get_away_stats('points')

The effect of covid-19 on EPL away teams points


count
effect                        
positive effect              8
relatively not affected      4
negative effect              3

The average away points before covid19 is 1.34
The average away points after covid19 is 1.54
The average stats changes of away points is 0.19
The average changes % of away points is 14.49 %


away_points_before  away_points_after  \
away_team                                                         
Liverpool                               2.45               1.79   
Crystal Palace                          1.39               1.00   
Wolverhampton Wanderers                 1.33               1.12   
Chelsea                                 1.64               1.67   
Southampton                             1.09               1.12   
Tottenham Hotspur                       1.41               1.48   
Manchester City                         2.12               2.22   
Newcastle United                        1.00               1.13   
Leicester City                          1.45               1.67   
Burnley                                 1.00               1.29   
Arsenal                                 1.22               1.56   
Brighton and Hove Albion                0.71               1.22   
West Ham United                         0.97               1.57   
Everton                                 0.94               1.87   
Manchester United                       1.39               2.33   

                          stats_changes  changes_%  
away_team                                           
Liverpool                         -0.66     -27.01  
Crystal Palace                    -0.39     -28.26  
Wolverhampton Wanderers           -0.21     -15.62  
Chelsea                            0.03       1.85  
Southampton                        0.03       3.13  
Tottenham Hotspur                  0.07       4.71  
Manchester City                    0.10       4.71  
Newcastle United                   0.13      13.04  
Leicester City                     0.21      14.58  
Burnley                            0.29      29.17  
Arsenal                            0.34      28.00  
Brighton and Hove Albion           0.51      72.46  
West Ham United                    0.59      61.26  
Everton                            0.93      98.64  
Manchester United                  0.94      67.39

Insight from Home/Away points per game :
* Majority of home teams suffered significantly in points per game department while away team generally have improved points per game.
* Southampton is the team with biggest home points per game increase after home crowd absence.
* Manchester United improved the most in away points per game.
* West Ham United is the only team that had improved both home/away point per game  with +0.39 and +0.94 in stats changes respectively.
* Liverpool suffered the most home/away point per game drop with -1.01 and -0.66 in stats changes respectively.
* Liverpool, Crystal Palace, Wolverhampton Wanderers are the teams that experienced decreased in both home/away point per game.
* Chelsea and Tottenham Hotspur are the only teams that is relatively not affected in both home/away point per game.

### Performance Stats

#### Home / Away Ball Possession %

In [49]:
get_home_stats('possession')

The effect of covid-19 on EPL home teams possession


count
effect                        
positive effect              3
relatively not affected      7
negative effect              5

The average home possession before covid19 is 53.11
The average home possession after covid19 is 52.43
The average stats changes of home possession is -0.68
The average changes % of home possession is -1.28 %


home_possession_before  home_possession_after  \
home_team                                                                 
Tottenham Hotspur                          58.73                  50.82   
Arsenal                                    59.00                  52.74   
Everton                                    52.55                  47.93   
Crystal Palace                             45.98                  41.80   
West Ham United                            47.63                  43.60   
Chelsea                                    62.71                  60.49   
Manchester City                            67.76                  66.72   
Leicester City                             55.70                  55.65   
Newcastle United                           41.60                  41.93   
Brighton and Hove Albion                   49.81                  51.43   
Liverpool                                  62.85                  64.78   
Burnley                                    42.55                  44.67   
Manchester United                          55.76                  58.85   
Wolverhampton Wanderers                    48.48                  52.39   
Southampton                                45.49                  52.62   

                          stats_changes  changes_%  
home_team                                           
Tottenham Hotspur                 -7.91     -13.47  
Arsenal                           -6.26     -10.60  
Everton                           -4.63      -8.80  
Crystal Palace                    -4.18      -9.09  
West Ham United                   -4.03      -8.45  
Chelsea                           -2.21      -3.53  
Manchester City                   -1.04      -1.54  
Leicester City                    -0.04      -0.08  
Newcastle United                   0.33       0.80  
Brighton and Hove Albion           1.61       3.24  
Liverpool                          1.93       3.06  
Burnley                            2.12       4.98  
Manchester United                  3.09       5.54  
Wolverhampton Wanderers            3.91       8.07  
Southampton                        7.13      15.67

In [50]:
get_away_stats('possession')

The effect of covid-19 on EPL away teams possession


count
effect                        
positive effect              4
relatively not affected      7
negative effect              4

The average away possession before covid19 is 50.21
The average away possession after covid19 is 50.26
The average stats changes of away possession is 0.05
The average changes % of away possession is 0.1 %


away_possession_before  away_possession_after  \
away_team                                                                 
Everton                                    49.26                  45.14   
Manchester City                            66.74                  62.74   
West Ham United                            45.55                  42.45   
Crystal Palace                             43.43                  40.45   
Arsenal                                    55.41                  52.71   
Tottenham Hotspur                          54.25                  51.94   
Liverpool                                  62.07                  61.16   
Brighton and Hove Albion                   46.15                  46.90   
Newcastle United                           35.99                  36.77   
Manchester United                          53.54                  54.75   
Chelsea                                    61.21                  62.85   
Burnley                                    38.42                  40.45   
Wolverhampton Wanderers                    45.12                  47.54   
Leicester City                             51.19                  55.57   
Southampton                                44.78                  52.43   

                          stats_changes  changes_%  
away_team                                           
Everton                           -4.12      -8.37  
Manchester City                   -4.00      -6.00  
West Ham United                   -3.10      -6.81  
Crystal Palace                    -2.98      -6.86  
Arsenal                           -2.69      -4.86  
Tottenham Hotspur                 -2.31      -4.26  
Liverpool                         -0.91      -1.46  
Brighton and Hove Albion           0.75       1.62  
Newcastle United                   0.78       2.18  
Manchester United                  1.22       2.27  
Chelsea                            1.64       2.67  
Burnley                            2.03       5.28  
Wolverhampton Wanderers            2.43       5.38  
Leicester City                     4.37       8.54  
Southampton                        7.65      17.08

Insight from Home/Away ball possession % :
* The teams are mostly unaffected in terms of home/away ball possession percentage.
* Southampton increased the most home/away ball possession % with +7.13 and +7.65 in stats changes respectively.
* Southampton & Wolverhampton Wanderers are the only two teams with increased home/away ball possession.
* Tottenham Hotspur experienced the most decrease in home ball possession % with -7.91 in stats changes. 
* Everton, Crystal Palace and West Ham United are the teams that experienced decreased in both home/away ball possession %.
* Chelsea, Newcastle United, Brighton and Hove Albion & Liverpool are the teams with relatively consistent home/away ball possession %.

#### Home/Away Goals Scored per Game

In [51]:
get_home_stats('score')

The effect of covid-19 on EPL home teams score


count
effect                        
positive effect              5
relatively not affected      4
negative effect              6

The average home score before covid19 is 1.66
The average home score after covid19 is 1.53
The average stats changes of home score is -0.13
The average changes % of home score is -7.74 %


home_score_before  home_score_after  stats_changes  \
home_team                                                                      
Liverpool                              2.79              1.78          -1.01   
Arsenal                                2.00              1.48          -0.52   
Burnley                                1.29              0.78          -0.51   
Everton                                1.48              1.21          -0.28   
Wolverhampton Wanderers                1.44              1.17          -0.27   
Manchester City                        2.81              2.68          -0.13   
Chelsea                                1.79              1.70          -0.10   
Brighton and Hove Albion               1.09              1.04          -0.05   
West Ham United                        1.67              1.62          -0.04   
Tottenham Hotspur                      1.85              1.83          -0.02   
Crystal Palace                         0.91              1.00           0.09   
Leicester City                         1.59              1.70           0.11   
Southampton                            1.26              1.43           0.17   
Manchester United                      1.82              2.13           0.31   
Newcastle United                       1.09              1.42           0.33   

                          changes_%  
home_team                            
Liverpool                    -36.20  
Arsenal                      -26.09  
Burnley                      -39.53  
Everton                      -18.62  
Wolverhampton Wanderers      -18.54  
Manchester City               -4.71  
Chelsea                       -5.49  
Brighton and Hove Albion      -4.51  
West Ham United               -2.50  
Tottenham Hotspur             -0.82  
Crystal Palace                 9.68  
Leicester City                 6.76  
Southampton                   13.45  
Manchester United             16.83  
Newcastle United              29.86

In [52]:
get_away_stats('score')

The effect of covid-19 on EPL away teams score


count
effect                        
positive effect              7
relatively not affected      3
negative effect              5

The average away score before covid19 is 1.36
The average away score after covid19 is 1.4
The average stats changes of away score is 0.04
The average changes % of away score is 3.04 %


away_score_before  away_score_after  stats_changes  \
away_team                                                                      
Crystal Palace                         1.39              0.96          -0.44   
Wolverhampton Wanderers                1.18              0.79          -0.39   
Everton                                1.24              1.09          -0.15   
Leicester City                         1.67              1.58          -0.08   
Chelsea                                1.61              1.54          -0.06   
Burnley                                1.06              1.00          -0.06   
Manchester City                        2.15              2.17           0.03   
Brighton and Hove Albion               0.91              0.96           0.04   
Arsenal                                1.41              1.48           0.07   
Tottenham Hotspur                      1.56              1.65           0.09   
Liverpool                              1.82              1.92           0.10   
Southampton                            1.12              1.25           0.13   
Newcastle United                       0.91              1.09           0.18   
Manchester United                      1.42              1.92           0.49   
West Ham United                        0.94              1.61           0.67   

                          changes_%  
away_team                            
Crystal Palace               -31.25  
Wolverhampton Wanderers      -33.01  
Everton                      -12.01  
Leicester City                -5.00  
Chelsea                       -4.01  
Burnley                       -5.71  
Manchester City                1.25  
Brighton and Hove Albion       4.91  
Arsenal                        5.24  
Tottenham Hotspur              5.99  
Liverpool                      5.42  
Southampton                   11.49  
Newcastle United              19.21  
Manchester United             34.57  
West Ham United               70.92

Insight from Home/Away goal scored per game :
* Generally, average home goal scored per game had declined while average away goal scored per game are relatively the same.
* Newcastle United increased the most in home goal scored per game with +0.33 in stats changes compared to before covid-19
* West Ham united is the team with most improved away goal scored per game with +0.67 in stats changes.
* Manchester United, Newcastle United & Southampton are the teams which increases in both home/away goal scored per game.
* Liverpool suffered the biggest lost in home goal scored per game with -1.01 in stats changes.
* Crystal Palace decreased 0.44 away goal scored per game, the biggest among all other away team.
* Wolverhampton Wanderers & Everton  are the teams that had worse stats in both home/away goal scored per game after covid-19.
* Manchester City & Brighton and Hove Albion are the teams that have relatively unchanged home/away goal scored per game.

#### Home/Away Shots on Target

In [53]:
get_home_stats('shots_on_target')

The effect of covid-19 on EPL home teams shots_on_target


count
effect                        
positive effect              3
relatively not affected      3
negative effect              9

The average home shots_on_target before covid19 is 5.02
The average home shots_on_target after covid19 is 4.7
The average stats changes of home shots_on_target is -0.32
The average changes % of home shots_on_target is -6.34 %


home_shots_on_target_before  \
home_team                                               
Everton                                          4.94   
West Ham United                                  4.88   
Liverpool                                        6.71   
Manchester United                                6.47   
Burnley                                          4.00   
Manchester City                                  7.38   
Arsenal                                          4.88   
Crystal Palace                                   3.82   
Leicester City                                   5.06   
Chelsea                                          6.21   
Newcastle United                                 4.06   
Tottenham Hotspur                                4.94   
Wolverhampton Wanderers                          4.18   
Southampton                                      4.38   
Brighton and Hove Albion                         3.33   

                          home_shots_on_target_after  stats_changes  changes_%  
home_team                                                                       
Everton                                         3.79          -1.15     -23.24  
West Ham United                                 4.04          -0.84     -17.16  
Liverpool                                       5.91          -0.79     -11.82  
Manchester United                               5.70          -0.77     -11.98  
Burnley                                         3.26          -0.74     -18.48  
Manchester City                                 6.68          -0.70      -9.42  
Arsenal                                         4.30          -0.58     -11.84  
Crystal Palace                                  3.26          -0.56     -14.72  
Leicester City                                  4.78          -0.28      -5.46  
Chelsea                                         6.13          -0.08      -1.22  
Newcastle United                                4.25           0.19       4.66  
Tottenham Hotspur                               5.17           0.23       4.60  
Wolverhampton Wanderers                         4.52           0.35       8.27  
Southampton                                     4.83           0.44      10.13  
Brighton and Hove Albion                        3.83           0.50      15.00

In [54]:
get_away_stats('shots_on_target')

The effect of covid-19 on EPL away teams shots_on_target


count
effect                        
positive effect              7
relatively not affected      2
negative effect              6

The average away shots_on_target before covid19 is 4.22
The average away shots_on_target after covid19 is 4.2
The average stats changes of away shots_on_target is -0.01
The average changes % of away shots_on_target is -0.34 %


away_shots_on_target_before  \
away_team                                               
Tottenham Hotspur                                4.59   
Manchester City                                  6.35   
Wolverhampton Wanderers                          3.88   
Everton                                          4.15   
Leicester City                                   5.03   
Brighton and Hove Albion                         3.50   
Crystal Palace                                   3.39   
Southampton                                      4.36   
Newcastle United                                 3.09   
Arsenal                                          3.53   
Liverpool                                        5.27   
Manchester United                                5.18   
Chelsea                                          4.73   
West Ham United                                  3.65   
Burnley                                          2.55   

                          away_shots_on_target_after  stats_changes  changes_%  
away_team                                                                       
Tottenham Hotspur                               3.57          -1.02     -22.30  
Manchester City                                 5.43          -0.92     -14.45  
Wolverhampton Wanderers                         3.33          -0.55     -14.06  
Everton                                         3.65          -0.49     -11.93  
Leicester City                                  4.67          -0.36      -7.23  
Brighton and Hove Albion                        3.17          -0.33      -9.32  
Crystal Palace                                  3.46           0.06       1.90  
Southampton                                     4.46           0.09       2.17  
Newcastle United                                3.30           0.22       7.00  
Arsenal                                         3.80           0.27       7.61  
Liverpool                                       5.54           0.27       5.10  
Manchester United                               5.50           0.32       6.14  
Chelsea                                         5.12           0.40       8.41  
West Ham United                                 4.48           0.83      22.79  
Burnley                                         3.54           1.00      39.14

Insight from Home/Away average shots on target :
* Generally, average home shots on target had dropped while average away shots on target were relatively unchanged.
* Newcastle United increased the most in shots on target per game with +0.50 in stats changes compared to before covid-19.
* Burnley is the team with most improved away shots on target per game with +1.00 in stats changes.
* Everton's home shots on target per game dropped the most with -1.15 in stats changes.
* Tottenham Hotspurs is the team with worst away shots on target per game changes with -1.02 in stats changes.
* Manchester City, Everton & Leicester are the teams with worse in both home/away shots on target per game after covid-19.
* There are no teams with both relatively unchanged home/away shots on target per game.

#### Home/Away Shots

In [55]:
get_home_stats('shots')

The effect of covid-19 on EPL home teams shots


count
effect                        
positive effect              1
relatively not affected      1
negative effect             13

The average home shots before covid19 is 14.56
The average home shots after covid19 is 13.06
The average stats changes of home shots is -1.5
The average changes % of home shots is -10.31 %


home_shots_before  home_shots_after  stats_changes  \
home_team                                                                      
Crystal Palace                        13.79              9.48          -4.32   
Everton                               14.61             11.08          -3.52   
Manchester City                       20.81             17.68          -3.13   
Tottenham Hotspur                     15.00             12.17          -2.83   
Chelsea                               17.29             15.30          -1.99   
Leicester City                        14.94             13.00          -1.94   
Burnley                               11.50             10.30          -1.20   
Manchester United                     15.62             14.43          -1.18   
Newcastle United                      13.36             12.21          -1.16   
Arsenal                               13.15             12.17          -0.97   
Liverpool                             17.35             16.39          -0.96   
West Ham United                       12.36             11.58          -0.78   
Southampton                           13.03             12.30          -0.73   
Wolverhampton Wanderers               13.53             13.70           0.17   
Brighton and Hove Albion              12.06             14.08           2.02   

                          changes_%  
home_team                            
Crystal Palace               -31.29  
Everton                      -24.12  
Manchester City              -15.05  
Tottenham Hotspur            -18.89  
Chelsea                      -11.51  
Leicester City               -12.99  
Burnley                      -10.40  
Manchester United             -7.57  
Newcastle United              -8.65  
Arsenal                       -7.40  
Liverpool                     -5.54  
West Ham United               -6.31  
Southampton                   -5.56  
Wolverhampton Wanderers        1.23  
Brighton and Hove Albion      16.77

In [56]:
get_away_stats('shots')

The effect of covid-19 on EPL away teams shots


count
effect                        
positive effect              5
relatively not affected      6
negative effect              4

The average away shots before covid19 is 11.66
The average away shots after covid19 is 11.67
The average stats changes of away shots is 0.01
The average changes % of away shots is 0.1 %


away_shots_before  away_shots_after  stats_changes  \
away_team                                                                      
Everton                               11.85              9.39          -2.46   
Wolverhampton Wanderers               11.79              9.88          -1.91   
Southampton                           12.45             11.21          -1.25   
Tottenham Hotspur                     11.59             10.57          -1.02   
Manchester City                       16.41             15.70          -0.72   
Chelsea                               15.15             14.50          -0.65   
Crystal Palace                         9.42              9.17          -0.26   
Newcastle United                       8.97              8.83          -0.14   
Brighton and Hove Albion              10.26             10.22          -0.05   
Manchester United                     12.85             12.88           0.03   
Arsenal                               10.38             10.92           0.54   
Leicester City                        12.48             13.38           0.89   
Burnley                                8.18              9.71           1.53   
Liverpool                             13.24             15.42           2.17   
West Ham United                        9.82             13.30           3.48   

                          changes_%  
away_team                            
Everton                      -20.77  
Wolverhampton Wanderers      -16.23  
Southampton                  -10.01  
Tottenham Hotspur             -8.83  
Manchester City               -4.36  
Chelsea                       -4.30  
Crystal Palace                -2.73  
Newcastle United              -1.61  
Brighton and Hove Albion      -0.46  
Manchester United              0.21  
Arsenal                        5.25  
Leicester City                 7.13  
Burnley                       18.66  
Liverpool                     16.42  
West Ham United               35.43

Insight from Home/Away average shots :
* Same as shots on target, average home shots had dropped while average away shots were relatively unchanged.
* Brighton and Hove Albion improved the most in number of home shots per game, increased 2.02 in stats changes.
* West Ham United increases the most in number of away shots per game with + 3.48 in stats changes.
* Crystal Palace is the team with most decrease in number of home shots per game with - 4.32 in stats changes.
* Everton's away shots per game dropped the most with -2.46 in stats changes.
* Everton, Southampton & Tottenham Hotspur become worse in both home/away number of shots per game.
* There are no teams with both relatively unchanged home/away shots per game.

#### Home/Away Shot on Target % 

In [57]:
get_home_stats('SoT%')

The effect of covid-19 on EPL home teams SoT%


count
effect                        
positive effect              8
relatively not affected      5
negative effect              2

The average home SoT% before covid19 is 34.94
The average home SoT% after covid19 is 37.04
The average stats changes of home SoT% is 2.1
The average changes % of home SoT% is 6.0 %


home_SoT%_before  home_SoT%_after  stats_changes  \
home_team                                                                    
Manchester United                    44.37            37.55          -6.82   
Burnley                              34.64            31.61          -3.03   
Liverpool                            39.05            37.29          -1.76   
West Ham United                      39.46            37.92          -1.54   
Everton                              34.06            33.68          -0.38   
Brighton and Hove Albion             27.46            28.17           0.71   
Leicester City                       35.16            36.32           1.16   
Arsenal                              37.01            39.07           2.06   
Wolverhampton Wanderers              31.02            33.19           2.17   
Manchester City                      35.74            38.94           3.20   
Southampton                          36.56            40.81           4.25   
Chelsea                              36.39            41.24           4.84   
Newcastle United                     30.58            36.46           5.88   
Tottenham Hotspur                    35.27            42.51           7.25   
Crystal Palace                       27.36            40.80          13.45   

                          changes_%  
home_team                            
Manchester United            -15.36  
Burnley                       -8.74  
Liverpool                     -4.51  
West Ham United               -3.90  
Everton                       -1.13  
Brighton and Hove Albion       2.58  
Leicester City                 3.29  
Arsenal                        5.56  
Wolverhampton Wanderers        7.00  
Manchester City                8.95  
Southampton                   11.62  
Chelsea                       13.30  
Newcastle United              19.23  
Tottenham Hotspur             20.55  
Crystal Palace                49.14

In [58]:
get_away_stats('SoT%')

The effect of covid-19 on EPL away teams SoT%


count
effect                        
positive effect              6
relatively not affected      3
negative effect              6

The average away SoT% before covid19 is 36.51
The average away SoT% after covid19 is 36.54
The average stats changes of away SoT% is 0.03
The average changes % of away SoT% is 0.08 %


away_SoT%_before  away_SoT%_after  stats_changes  \
away_team                                                                    
Tottenham Hotspur                    40.65            34.87          -5.78   
West Ham United                      37.73            33.06          -4.68   
Liverpool                            39.90            36.62          -3.29   
Manchester City                      40.14            37.85          -2.30   
Arsenal                              36.90            34.64          -2.26   
Crystal Palace                       37.69            35.48          -2.22   
Leicester City                       39.59            38.06          -1.53   
Brighton and Hove Albion             33.80            33.08          -0.72   
Wolverhampton Wanderers              32.64            32.74           0.10   
Southampton                          37.03            40.20           3.17   
Manchester United                    39.96            43.61           3.65   
Newcastle United                     33.48            37.22           3.74   
Chelsea                              32.08            35.87           3.78   
Everton                              35.13            39.46           4.34   
Burnley                              30.85            35.28           4.43   

                          changes_%  
away_team                            
Tottenham Hotspur            -14.22  
West Ham United              -12.39  
Liverpool                     -8.24  
Manchester City               -5.72  
Arsenal                       -6.12  
Crystal Palace                -5.88  
Leicester City                -3.86  
Brighton and Hove Albion      -2.12  
Wolverhampton Wanderers        0.31  
Southampton                    8.56  
Manchester United              9.14  
Newcastle United              11.17  
Chelsea                       11.79  
Everton                       12.35  
Burnley                       14.37

Insight from Home/Away shot on target percentage :
* Counterintuitively, home shot on target % had increased while away shots on target % stays relatively the same.
* Crystal Palace improved the most in home shot on target % with + 13.45 in stats changes.
* Burnley is the team with biggest improvement in away shot on target % with 4.43 in stats changes.
* Chelsea, Newcastle & Southampton are the teams with both improved home/away shot on target %.
* Manchester United is the team which suffered the most decreases in home shot on target %, with - 6.82 in stats changes.
* Tottenham Hotspur decreases the most in away shot on target % with - 5.78 in stats changes.
* There are no teams with decreased in both home/away shot on target percentage.
* Leicester City & Brighton and Hove Albion are the only teams that have relatively unchanged home/away shot on target %.

#### Home/Away Quantity Conversion Rate

In [59]:
get_home_stats('quan_CR')

The effect of covid-19 on EPL home teams quan_CR


count
effect                        
positive effect              9
relatively not affected      1
negative effect              5

The average home quan_CR before covid19 is 12.04
The average home quan_CR after covid19 is 13.21
The average stats changes of home quan_CR is 1.16
The average changes % of home quan_CR is 9.66 %


home_quan_CR_before  home_quan_CR_after  \
home_team                                                           
Liverpool                               16.65               12.62   
Wolverhampton Wanderers                 12.47                8.64   
Burnley                                 11.96                8.14   
Brighton and Hove Albion                10.58                8.50   
Arsenal                                 15.34               14.33   
Chelsea                                 10.69               11.13   
Manchester United                       13.12               13.98   
Leicester City                          11.52               12.54   
Everton                                 10.22               11.94   
Manchester City                         14.21               16.17   
Southampton                             10.54               12.73   
Tottenham Hotspur                       14.61               17.66   
Newcastle United                         9.47               13.46   
West Ham United                         12.68               17.62   
Crystal Palace                           6.59               18.67   

                          stats_changes  changes_%  
home_team                                           
Liverpool                         -4.03     -24.20  
Wolverhampton Wanderers           -3.84     -30.77  
Burnley                           -3.82     -31.96  
Brighton and Hove Albion          -2.08     -19.67  
Arsenal                           -1.01      -6.58  
Chelsea                            0.44       4.13  
Manchester United                  0.86       6.52  
Leicester City                     1.01       8.80  
Everton                            1.71      16.75  
Manchester City                    1.95      13.75  
Southampton                        2.19      20.79  
Tottenham Hotspur                  3.04      20.84  
Newcastle United                   3.99      42.12  
West Ham United                    4.94      38.93  
Crystal Palace                    12.09     183.57

In [60]:
get_away_stats('quan_CR')

The effect of covid-19 on EPL away teams quan_CR


count
effect                        
positive effect              7
relatively not affected      3
negative effect              5

The average away quan_CR before covid19 is 12.75
The average away quan_CR after covid19 is 12.28
The average stats changes of away quan_CR is -0.47
The average changes % of away quan_CR is -3.69 %


away_quan_CR_before  away_quan_CR_after  \
away_team                                                           
Burnley                                 16.21                9.09   
Crystal Palace                          17.15               11.49   
Wolverhampton Wanderers                 11.97                7.75   
Arsenal                                 16.29               14.22   
Liverpool                               14.07               12.55   
Chelsea                                 10.90               10.99   
Newcastle United                        10.09               10.36   
Tottenham Hotspur                       14.91               15.44   
Brighton and Hove Albion                 9.82               10.42   
Leicester City                          13.87               14.79   
Southampton                             10.14               11.13   
Manchester City                         13.88               15.65   
West Ham United                          9.65               11.47   
Everton                                 10.49               13.05   
Manchester United                       11.76               15.73   

                          stats_changes  changes_%  
away_team                                           
Burnley                           -7.11     -43.88  
Crystal Palace                    -5.66     -32.98  
Wolverhampton Wanderers           -4.22     -35.25  
Arsenal                           -2.08     -12.75  
Liverpool                         -1.51     -10.75  
Chelsea                            0.09       0.81  
Newcastle United                   0.27       2.68  
Tottenham Hotspur                  0.53       3.55  
Brighton and Hove Albion           0.60       6.11  
Leicester City                     0.92       6.66  
Southampton                        0.99       9.80  
Manchester City                    1.77      12.74  
West Ham United                    1.82      18.87  
Everton                            2.55      24.35  
Manchester United                  3.97      33.72

Insight from Home/Away Quantity Conversion Rate Percentage :
* Same as shot on target percentage, quantity conversion rate percentage also increased for home team while relatively unchanged for away team.
* Crystal Palace improved the most in home quantity conversion rate percentage with + 12.09 in stats changes.
* Manchester United is the team with most improved away quantity conversion rate percentage with + 3.97 in stats changes.
* Manchester United, Everton, West Ham United, Manchester City, Southampton &  Leicester City are the teams with both improved home/away quantity conversion rate percentage.
* Liverpool became worse in home quantity conversion rate percentage with - 4.03 conversion rate after covid-19.
* Burnley decreases the most in away quantity conversion rate with - 7.11 in stats changes.
* Burnley, Wolverhampton Wanderers, Arsenal & Liverpool are teams with decreased in both home/away quantity conversion rate.
* Chelsea are the only team with relatively unchanged quantity conversion rate.

#### Home/Away Quality Conversion Rate

In [61]:
get_home_stats('qual_CR')

The effect of covid-19 on EPL home teams qual_CR


count
effect                        
positive effect              7
relatively not affected      2
negative effect              6

The average home qual_CR before covid19 is 34.59
The average home qual_CR after covid19 is 33.9
The average stats changes of home qual_CR is -0.7
The average changes % of home qual_CR is -2.01 %


home_qual_CR_before  home_qual_CR_after  \
home_team                                                           
Liverpool                               43.14               29.11   
Burnley                                 40.39               26.85   
Wolverhampton Wanderers                 38.27               27.72   
Arsenal                                 42.15               32.54   
Brighton and Hove Albion                35.00               27.39   
Chelsea                                 32.38               27.25   
Southampton                             30.21               30.14   
Leicester City                          32.78               33.84   
Newcastle United                        29.96               31.94   
Manchester City                         38.18               40.66   
Tottenham Hotspur                       37.67               41.83   
Everton                                 32.83               40.60   
Manchester United                       27.96               35.87   
West Ham United                         31.87               43.37   
Crystal Palace                          26.10               39.36   

                          stats_changes  changes_%  
home_team                                           
Liverpool                        -14.03     -32.53  
Burnley                          -13.54     -33.53  
Wolverhampton Wanderers          -10.55     -27.56  
Arsenal                           -9.62     -22.81  
Brighton and Hove Albion          -7.61     -21.74  
Chelsea                           -5.13     -15.84  
Southampton                       -0.07      -0.22  
Leicester City                     1.06       3.22  
Newcastle United                   1.98       6.60  
Manchester City                    2.48       6.49  
Tottenham Hotspur                  4.16      11.05  
Everton                            7.76      23.65  
Manchester United                  7.92      28.32  
West Ham United                   11.50      36.09  
Crystal Palace                    13.25      50.78

In [62]:
get_away_stats('qual_CR')

The effect of covid-19 on EPL away teams qual_CR


count
effect                        
positive effect              7
relatively not affected      1
negative effect              7

The average away qual_CR before covid19 is 33.61
The average away qual_CR after covid19 is 32.55
The average stats changes of away qual_CR is -1.06
The average changes % of away qual_CR is -3.15 %


away_qual_CR_before  away_qual_CR_after  \
away_team                                                           
Burnley                                 46.23               25.40   
Crystal Palace                          43.67               32.61   
Wolverhampton Wanderers                 34.14               24.06   
Newcastle United                        34.14               29.38   
Chelsea                                 35.21               30.73   
Arsenal                                 40.90               37.66   
Liverpool                               34.80               32.88   
Leicester City                          34.25               35.37   
Everton                                 28.04               29.83   
Southampton                             25.36               28.67   
Manchester United                       29.93               34.70   
Tottenham Hotspur                       37.38               43.03   
Brighton and Hove Albion                24.21               30.07   
Manchester City                         32.87               40.71   
West Ham United                         22.98               33.11   

                          stats_changes  changes_%  
away_team                                           
Burnley                          -20.83     -45.06  
Crystal Palace                   -11.06     -25.33  
Wolverhampton Wanderers          -10.08     -29.53  
Newcastle United                  -4.76     -13.95  
Chelsea                           -4.48     -12.72  
Arsenal                           -3.24      -7.93  
Liverpool                         -1.92      -5.51  
Leicester City                     1.12       3.28  
Everton                            1.80       6.40  
Southampton                        3.31      13.07  
Manchester United                  4.77      15.95  
Tottenham Hotspur                  5.65      15.12  
Brighton and Hove Albion           5.86      24.21  
Manchester City                    7.84      23.86  
West Ham United                   10.13      44.06

Insight from Home/Away Quality Conversion Rate Percentage :
* Quality conversion rate percentage stays relatively unchanged in both home/away despite increased home shot on target percentage.
* Crystal Palace improved the most in home quality conversion rate percentage with + 13.25 in stats changes.
* West Ham United is the team with most improved away quality conversion rate percentage with +0.10 in stats changes.
* West Ham United, Manchester City, Tottenham Hotspur, Manchester United & Everton are the teams with both increased home/away quality conversion rate percentage.
* Liverpool became worse in home quality conversion rate percentage with - 14.03 conversion rate after covid-19.
* Burnley decreases the most in away quantity conversion rate percentage with - 20.83 in stats changes.
* Burnley, Wolverhampton Wanderers, Chelsea, Arsenal & Liverpool are teams which became worse in both home/away quality conversion rate percentage.
* Leicester City are the only team with relatively unchanged quality conversion rate percentage.

## Ball Possession vs Conversion Rate

It can show how ball possession affect ability to convert shots & shots on target into goal

In [53]:
#filter home data
home_data = epl_data.loc[(epl_data['home_team'].isin(filtered_teams))]

#filter away data 
away_data = epl_data.loc[(epl_data['away_team'].isin(filtered_teams))]

In [48]:
import plotly.express as px

In [73]:
home_possession_vs_quanCR = px.scatter(home_data, x='home_quan_CR', y='home_possession',color='covid19')
home_possession_vs_quanCR.show()

In [82]:
home_possession_vs_qualCR = px.scatter(home_data, x='home_qual_CR', y='home_possession',color='covid19')
home_possession_vs_qualCR.show()

In [84]:
away_possession_vs_quanCR = px.scatter(away_data, x='away_quan_CR', y='away_possession',color='covid19')
away_possession_vs_quanCR.show()

In [83]:
away_possession_vs_qualCR = px.scatter(away_data, x='away_qual_CR', y='away_possession',color='covid19')
away_possession_vs_qualCR.show()

## Possession vs Score

It can show relation between ball possession and goal-scoring ability 

In [71]:
home_possession_vs_score = px.box(home_data, x='home_score', y='home_possession',color='covid19')
home_possession_vs_score.show()

In [74]:
away_possession_vs_score = px.box(home_data, x='away_score', y='away_possession',color='covid19')
away_possession_vs_score.show()

## Shot on Target % vs Score

It can show how shot accuracy can affect goal scoring

In [66]:
home_SoTpct_vs_score = px.box(home_data, x='home_score', y='home_SoT%',color='covid19')
home_SoTpct_vs_score.show()

In [80]:
away_SoTpct_vs_score = px.box(away_data, x='away_score', y='away_SoT%',color='covid19')
away_SoTpct_vs_score.show()

## Shot on Target % vs Point

It can show how performance translates into result

In [70]:
home_SoTpct_vs_point = px.box(home_data, x='home_points', y='home_SoT%',color='covid19')
home_SoTpct_vs_point.show()

In [81]:
away_SoTpct_vs_point = px.box(away_data, x='away_points', y='away_SoT%',color='covid19')
away_SoTpct_vs_point.show()

## Big 6 vs Non Big 6